<h3>
    Introduction
</h3>
<p>
    Pathway analyses are done using fold-change values of all genes calculated
    by <em>limma</em> or <em>DESeq2</em>. This is different from the enrichment
    analysis in the DEG2 tab, which only uses gene lists of differentially
    expressed genes (DEGs).
</p>
<p>
    Our method also differs from traditional methods like
    <a href="https://www.gsea-msigdb.org/gsea/index.jsp">
        Gene Set Enrichment Analysis (GSEA),
    </a>
    which takes the normalized expression matrix and conducts more
    sophisticated analyses like permutation tests. For confirmation, please use
    the stand-alone GSEA.
</p>
<h3>
    Methods
</h3>
<p>
    Pathway analysis can be performed using several methods. GSEA (Gene Set
    Enrichment Analysis) (Subramanian et al., 2005) is conducted in the
    pre-ranked mode using a recent faster algorithm based on the <em>fgsea</em>
    package (Sergushichev, 2016). PAGE (Parametric Analysis of Gene Set
    Enrichment) (Kim and Volsky, 2005) is used as implemented in <em>PGSEA</em>
    package (Furge and Dykema, 2012). For PGSEA there are two versions. One
    only analyzes the selected comparisons and another option (“PGSEA w/ all
    samples”) enables the user to analyze all sample groups.
</p>
<p>
    Unlike all these methods that rely on the built-in geneset databases,
    ReactomePA (Reactome Pathway Analysis) (Yu and He, 2016) retrieves genesets
    from Reactome (Fabregat et al., 2016; Yu and He, 2016).
</p>
<p>
    Gene expression data can be visualized on KEGG pathway diagrams (Kanehisa
    et al., 2017) using Pathview (Luo and Brouwer, 2013). Note that Pathview
    download pathway diagrams directly from KEGG website and thus is slow.
</p>
<p>
    On the lower left side of the screen, there is a checkbox named “Use
    absolute values of fold changes for GSEA and GAGE”. This is useful as some
    molecular pathways can be modulated by up-regulating some genes while
    down-regulating others. This is especially useful when using KEGG pathways.
    For other genes sets such as TF target genes and microRNA target genes
    where we know the regulation is one-directional, we should not check this
    box.
</p>
<p>
    Users can choose to filter out some genes with noisy fold-changes by using
    an FDR cutoff by reducing the “Remove genes with big FDR before pathway
    analysis” from the default value of 1 to a relatively bigger FDR cutoff
    like 0.6. If this is done, genes with FDR &gt; 0.6 is removed from pathway
    analysis.
</p>
<h3>
    Interpretation
</h3>
<p>
    iDEP makes it easy for users to choose different pathway analysis methods
    and databases. The drawback is that it is easy to rationalize any a priori
    hypothesis (bias). With so many choices and knobs to turn, you can make any
    story you wanted to make!
</p>
<p>
    Results from pathway analysis do not directly inform about the activity of
    the pathway of interest but often reflect downstream effects of changes in
    the activities of transcription factors.
</p>
<h3>
    Pathway Database
</h3>
<p>
    Very small genesets can cause problems. By default, genesets with less than
    15 genes are disregarded. Sometimes this needs to be reduced to 10, or even
    5, when some pathways of interest only has few genes. But be aware that
    this may introduce false positives.
</p>
<p>
    Besides Gene Ontology, iDEP includes additional data from KEGG, Reactome,
    and many other sources. There might be errors in our conversion of these
    diverse databases.
</p>
<h3>
    R Code
</h3>
<p>
    R code used for GAGE using the gage package:
</p>
<p>
    paths &lt;- gage(fold, gsets = gmt, ref = NULL, samp = NULL)
</p>
<p>
    R code for GSEA via fgsea pckage:
</p>
<p>
    paths &lt;- fgsea(pathways = gmt,
</p>
<p>
    stats = fold,
</p>
<p>
    minSize=input$minSetSize,
</p>
<p>
    maxSize=input$maxSetSize,
</p>
<p>
    nperm=5000)
</p>
<p>
    R code for PAGE using the PGSEA package:
</p>
<p>
    pg= PGSEA (convertedData – rowMeans(convertedData), cl=gmt, range=myrange,
    p.value=TRUE, weighted=FALSE)
</p>
<p>
    R code for ReactomePA:
</p>
<p>
    paths &lt;- gsePathway(fold, nPerm=50000, organism = ReactomePASpecies[ix],
</p>
<p>
    minGSSize= input$minSetSize,
</p>
<p>
    maxGSSize= input$maxSetSize,
</p>
<p>
    pvalueCutoff=0.5,
</p>
<p>
    pAdjustMethod=”BH”, verbose=FALSE)
</p>
<h3>
    References
</h3>
<p>
    Fabregat, A., Sidiropoulos, K., Garapati, P., Gillespie, M., Hausmann, K.,
    Haw, R., Jassal, B., Jupe, S., Korninger, F., McKay, S., et al. (2016). The
    Reactome pathway Knowledgebase. Nucleic Acids Res 44, D481-487.
</p>
<p>
    Furge, K., and Dykema, K. (2012). PGSEA: Parametric Gene Set Enrichment
    Analysis. R package version 1480.
</p>
<p>
    Kanehisa, M., Furumichi, M., Tanabe, M., Sato, Y., and Morishima, K.
    (2017). KEGG: new perspectives on genomes, pathways, diseases and drugs.
    Nucleic Acids Res 45, D353-D361.
</p>
<p>
    Kim, S.Y., and Volsky, D.J. (2005). PAGE: parametric analysis of gene set
    enrichment. BMC Bioinformatics 6, 144.
</p>
<p>
    Luo, W., and Brouwer, C. (2013). Pathview: an R/Bioconductor package for
    pathway-based data integration and visualization. Bioinformatics 29,
    1830-1831.
</p>
<p>
    Sergushichev, A. (2016). An algorithm for fast preranked gene set
    enrichment analysis using cumulative statistic calculation. bioRxiv
    http://biorxiv.org/content/early/2016/06/20/060012.
</p>
<p>
    Subramanian, A., Tamayo, P., Mootha, V.K., Mukherjee, S., Ebert, B.L.,
    Gillette, M.A., Paulovich, A., Pomeroy, S.L., Golub, T.R., Lander, E.S., et
    al. (2005). Gene set enrichment analysis: a knowledge-based approach for
    interpreting genome-wide expression profiles. Proc Natl Acad Sci U S A 102,
    15545-15550.
</p>
<p>
    Yu, G., and He, Q.Y. (2016). ReactomePA: an R/Bioconductor package for
    Reactome pathway analysis and visualization. Mol Biosyst 12, 477-479.
</p>